### PR TITLE
[HttpKernel] Added support for registering Bundle dependencies

### PR DESCRIPTION
**Pull request by andrerom**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes https://github.com/symfony/symfony/pull/13945 |
| Applies to | 2.7 |
| Fixed tickets |  |

Todo:
- [ ] Update for Optional dependencies as just added to code PR

### CODE REVIEW

File: cookbook/bundles/dependencies.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+If your bundle is meant for reuse, then this will typically complicate your
+installation documentation, and if your dependencies change over time they
+will likewise complicate your upgrade documentation, and make upgrading
+more tedious for your users.
+
+Starting with Symfony 2.7 you can avoid this by specifying your dependencies
+by implementing "BundleDependenciesInterface" interface::
+
+    // src/Acme/UserBundle/AcmeCacheBundle.php
+    namespace Acme\CacheBundle;
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: please remove the vendor

File: cookbook/bundles/dependencies.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    // src/Acme/UserBundle/AcmeFileBundle.php
+    namespace Acme\FileBundle;
+
+    use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+    class AcmeFileBundle extends Bundle implements BundleDependenciesInterface
+    {
+        public function getBundleDependencies()
+        {
+            return array(Oneup\FlysystemBundle\OneupFlysystemBundle::CLASS);
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: Can we make this section a bit shorter and inside a tip block? (it's not really feature specific, but still a very nice tip)

```rst
.. tip::

    If your bundle requires PHP 5.5 or higher, you can also take advantage of
    the ``CLASS`` constant::

        use FOS\HttpCacheBundle\FOSHttpCacheBundle;

        // ...
        public function getBundleDependencies()
        {
            return array(FOSHttpCacheBundle::CLASS);
        }
```

File: cookbook/bundles/dependencies.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        }
+    }
+
+Either way you can provide as many strings as you want, as long as the format
+is Fully Qualified Name of the Bundle Class, in format as shown above.
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: I don't think this is required

File: cookbook/bundles/dependencies.rst (hunk 1 above)
Comment (andrerom) [inline cookbook/bundles/dependencies.rst]: Ok, note that this is the case in [inheritance.rst](https://github.com/andrerom/symfony-docs/blob/patch-1/cookbook/bundles/inheritance.rst)

File: cookbook/bundles/dependencies.rst [hunk 4]
```diff
@@ -0,0 +1,51 @@
+.. index::
+   single: Bundle; Dependencies
+
+How to Use Bundle Dependencies to load other Bundles
+====================================================
+
+When working on your own bundle(s), you'll sometimes have the need to reuse
+other bundles, either by just requiring, overriding or inherit from them.
+
+While Composer takes care about letting you make sure these dependencies are
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: "[...] takes care about making sure [...]"

File: cookbook/bundles/dependencies.rst [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+loaded, you'll also need to enable these bundles in the Kernel.
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: I would just lowercase "kernel"

File: cookbook/bundles/dependencies.rst (hunk 2 above)
Comment (andrerom) [inline cookbook/bundles/dependencies.rst]: What about the way it was done in the api doc?:

```
     * Example of use:
     * ```php
     * class AcmeBundle extends Bundle implements BundleDependenciesInterface
     * {
     *     public function getBundleDependencies()
     *     {
     *         return array(
     *
     *             // All values must be FQN strings to avoid bundles being loaded several times
     *             'FOS\HttpCacheBundle\FOSHttpCacheBundle',
     *
     *             // If you require PHP 5.5 or higher it is better to use `::CLASS` constant:
     *             Oneup\FlysystemBundle\OneupFlysystemBundle::CLASS
     *         );
     *     }
     * }
     * ```
```

File: cookbook/bundles/dependencies.rst [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+If your bundle is meant for reuse, then this will typically complicate your
+installation documentation, and if your dependencies change over time they
+will likewise complicate your upgrade documentation, and make upgrading
+more tedious for your users.
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: this is a very long sentence. What do you think of breaking it up like: "If your bundle is meant to be reused, bundle dependencies will complicate your installation documents. This makes installation and upgrading your bundlemore tedious for your users."

File: cookbook/bundles/dependencies.rst [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+Starting with Symfony 2.7 you can avoid this by specifying your dependencies
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]:
```rst
You can avoid this by specifying your dependencies by implementing the
:class:`Symfony\\Component\\HttpKernel\\Bundle\\BundleDependenciesInterface`::
```

Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: and then add a versionadded after the article title:

```rst
.. versionadded:: 2.7
    Bundle dependencies was introduced in Symfony 2.7.
```

File: cookbook/bundles/dependencies.rst (hunk 6 above)
Comment (andrerom) [inline cookbook/bundles/dependencies.rst]: good point

what about _"If your bundle is meant to be reused, bundle dependencies will complicate your installation documentation. This makes installation and upgrading your bundle more tedious for your users."_

**Comment by WouterJ** [comment]
Thank you for writing this nice article, @andrerom! I left 2 more comments (you're quick with the fixes!). You also have to add this article to `cookbook/map.rst.inc`, to make it show up on the site.

**Comment by andrerom** [comment]
Thanks for the review, I have updated the things you pointed out now, but might be I misunderstood on things like removing vendor or the :'class: syntax for instance. But should be ready for second look now.

File: cookbook/bundles/dependencies.rst [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+While Composer takes care about making sure these dependencies are loaded,
+you'll also need to enable these bundles in the kernel.
+
+If your bundle is meant to be reused, bundle dependencies will complicate
+your installation documentation. This makes installation and upgrading your
+bundle more tedious for your users.
+
+You can avoid this by specifying your dependencies, this will make sure
+they are loaded, and *before* your own bundle to make sure you can extend
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: "You can avoid this by specifying your dependencies. This will make sure that they are loaded in the kernel. It'll also make sure they are loaded *before* your own bundle, to make sure you can extend the bundle." (other than extending a bundle, it doesn't have to be loaded before afaik)

File: cookbook/bundles/dependencies.rst [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+or customize its configuration if needed.
+
+Specifying dependencies is accomplished by implementing
+:class:`Symfony\\Component\\HttpKernel\\Bundle\\BundleDependenciesInterface`::::
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: you should remove 2 colons at the end of the line here

File: cookbook/bundles/dependencies.rst [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+    // src/CacheBundle/AcmeCacheBundle.php
+    namespace CacheBundle;
+
+    use Symfony\Component\HttpKernel\Bundle\Bundle;
+    use Symfony\Component\HttpKernel\Bundle\BundleDependenciesInterface;
+
+    class AcmeCacheBundle extends Bundle implements BundleDependenciesInterface
```
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: The vendor name should also be removed from the bundle name. So you get `CacheBundle\CacheBundle`

**Comment by WouterJ** [comment]
Apart from some minor things :+1:

File: cookbook/bundles/dependencies.rst (hunk 8 above)
Comment (andrerom) [inline cookbook/bundles/dependencies.rst]: fixed, with one twist. I did not end the sentence with "bundle", but instead "them" as the subject is the dependencies so that felt clearer, might be even more clearer by consistently saying *the dependencies".  ok?

**Comment by andrerom** [comment]
Updated

File: cookbook/bundles/dependencies.rst [hunk 11]
```diff
@@ -0,0 +1,52 @@
+.. index::
+   single: Bundle; Dependencies
+
+How to Use Bundle Dependencies to load other Bundles
+====================================================
+
+.. versionadded:: 2.7
+    Bundle dependencies was introduced in Symfony 2.7.
```
Comment (xabbuh) [inline cookbook/bundles/dependencies.rst]: were

File: cookbook/bundles/dependencies.rst [hunk 12]
```diff
@@ -0,0 +1,52 @@
+.. index::
+   single: Bundle; Dependencies
+
+How to Use Bundle Dependencies to load other Bundles
```
Comment (xabbuh) [inline cookbook/bundles/dependencies.rst]: Load

File: cookbook/bundles/dependencies.rst [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+====================================================
+
+.. versionadded:: 2.7
+    Bundle dependencies was introduced in Symfony 2.7.
+
+When working on your own bundle(s), you'll sometimes have the need to reuse
+other bundles, either by just requiring, overriding or inherit from them.
```
Comment (xabbuh) [inline cookbook/bundles/dependencies.rst]: "inheriting" to be consistent with "requiring" and "overriding"?

File: cookbook/bundles/dependencies.rst [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+While Composer takes care about making sure these dependencies are loaded,
+you'll also need to enable these bundles in the kernel.
+
+If your bundle is meant to be reused, bundle dependencies will complicate
+your installation documentation. This makes installation and upgrading your
+bundle more tedious for your users.
+
+You can avoid this by specifying your dependencies. This will make sure that
+they are loaded in the kernel. It'll also make sure they are loaded *before*
+your own bundle, to make sure you can extend the them.
```
Comment (xabbuh) [inline cookbook/bundles/dependencies.rst]: [...] extend them.

File: cookbook/bundles/dependencies.rst [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+Specifying dependencies is accomplished by implementing
```
Comment (xabbuh) [inline cookbook/bundles/dependencies.rst]: [...] implementing the [...]

File: cookbook/bundles/dependencies.rst (hunk 11 above)
Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: Imo it's Was, as it's one feature.

Comment (xabbuh) [inline cookbook/bundles/dependencies.rst]: Then we should reword this to something like "Support for bundle dependencies was introduced [...]"

File: cookbook/bundles/dependencies.rst (hunk 12 above)
Comment (andrerom) [inline cookbook/bundles/dependencies.rst]: uppercase *L* ??

Comment (WouterJ) [inline cookbook/bundles/dependencies.rst]: Our titles case standard is to capitialize all words, except from closed-class words. Load is not closed-class (you can say "quickly load" for instance) and has to be capitialized

Comment (andrerom) [inline cookbook/bundles/dependencies.rst]: ok, fixed.

**Comment by andrerom** [comment]
Closed, updated,  and moved to new PR for 2.8: #5610
